ComponentKit/Debug/CKComponentDebugController.mm

```diff
@@ -133,46 +132,35 @@ + (BOOL)debugMode
 
 #pragma mark - Synchronous Reflow
 
+static NSHashTable<id<CKComponentDebugReflowListener>> *reflowListeners;
+static std::mutex reflowMutex;
+
 + (void)reflowComponents
 {
   if (![NSThread isMainThread]) {
     dispatch_async(dispatch_get_main_queue(), ^{ [self reflowComponents]; });
-  } else {
-    UIWindow *window = [[UIApplication sharedApplication] keyWindow];
-    CKRecursiveComponentReflow(window);
+    return;
   }
-}
-
-+ (void)reflowComponentsForView:(UIView *)view searchUpwards:(BOOL)upwards
-{
-  if (upwards) {
-    while (view && ![view isKindOfClass:[CKComponentRootView class]]) {
-      view = view.superview;
-    }
+  NSArray<id<CKComponentDebugReflowListener>> *copiedListeners;
+  {
+    std::lock_guard<std::mutex> l(reflowMutex);
+    copiedListeners = [reflowListeners allObjects];
   }
-  if (view) {
-    CKRecursiveComponentReflow(view);
+  for (id<CKComponentDebugReflowListener> listener in copiedListeners) {
+    [listener didReceiveReflowComponentsRequest];
   }
 }
 
-static void CKRecursiveComponentReflow(UIView *view)
++ (void)registerReflowListener:(id<CKComponentDebugReflowListener>)listener
 {
-  if (view.ck_componentLifecycleManager) {
-    CKComponentLifecycleManager *lifecycleManager = view.ck_componentLifecycleManager;
-    CKComponentLifecycleManagerState oldState = [lifecycleManager state];
-    CKComponentLifecycleManagerState state =
-    [lifecycleManager prepareForUpdateWithModel:oldState.model
-                                constrainedSize:oldState.constrainedSize
-                                        context:oldState.context];
-    [lifecycleManager updateWithState:state];
-  } else if ([view.superview isKindOfClass:[CKComponentHostingView class]]) {
-    CKComponentHostingView *hostingView = (CKComponentHostingView *)view.superview;
-    [hostingView setNeedsLayout];
-  } else {
-    for (UIView *subview in view.subviews) {
-      CKRecursiveComponentReflow(subview);
-    }
+  if (listener == nil) {
+    return;
+  }
+  std::lock_guard<std::mutex> l(reflowMutex);
+  if (reflowListeners == nil) {
+    reflowListeners = [NSHashTable weakObjectsHashTable];
```

Comment:
Nice. So we weakly hold the reflow listeners to naturally keep the collection bounded.